src/multisig/signing_protocol.cpp

```diff
@@ -0,0 +1,967 @@
+  cryptonote::transaction& unsigned_tx,
+  std::vector<cached_CLSAG_Gen_t>& cached_CLSAG,
+  rct::keyV& cached_w
+)
+{
+  if (rct_config.bp_version != 3)
+    return false;
+  if (rct_config.range_proof_type != rct::RangeProofPaddedBulletproof)
+    return false;
+
+  const std::size_t num_destinations = destinations.size();
+  const std::size_t num_sources = sources.size();
+
+  // rct_signatures component of tx
+  rct::rctSig rv{};
+
+  // set misc. fields
+  rv.type = rct::RCTTypeCLSAG;
+  rv.txnFee = fee;
+  rv.message = rct::hash2rct(cryptonote::get_transaction_prefix_hash(unsigned_tx));
+
+  // define outputs
+  std::vector<std::uint64_t> output_amounts(num_destinations);
+  rct::keyV output_amount_masks(num_destinations);
+  rv.ecdhInfo.resize(num_destinations);
+  rv.outPk.resize(num_destinations);
+  for (std::size_t i = 0; i < num_destinations; ++i) {
+    rv.outPk[i].dest = output_public_keys[i];
+    output_amounts[i] = destinations[i].amount;
+    output_amount_masks[i] = genCommitmentMask(output_amount_secret_keys[i]);
+    rv.ecdhInfo[i].amount = rct::d2h(output_amounts[i]);
+    rct::addKeys2(
+      rv.outPk[i].mask,
+      output_amount_masks[i],
+      rv.ecdhInfo[i].amount,
+      rct::H
+    );
+    rct::ecdhEncode(rv.ecdhInfo[i], output_amount_secret_keys[i], true);
+  }
+
+  // output range proofs
+  if (not reconstruction) {
+    rv.p.bulletproofs.push_back(rct::bulletproof_PROVE(output_amounts, output_amount_masks));
+  }
+  else {
+    rv.p.bulletproofs = unsigned_tx.rct_signatures.p.bulletproofs;
+    if (rv.p.bulletproofs.size() != 1)
+      return false;
+    rv.p.bulletproofs[0].V.resize(num_destinations);
+    for (std::size_t i = 0; i < num_destinations; ++i) {
+      rv.p.bulletproofs[0].V[i] = rct::scalarmultKey(rv.outPk[i].mask, rct::INV_EIGHT);
+    }
+    if (not bulletproof_VERIFY(rv.p.bulletproofs))
+      return false;
+  }
+
+  // prepare rings for input CLSAGs
+  rv.mixRing.resize(num_sources);
+  for (std::size_t i = 0; i < num_sources; ++i) {
+    const std::size_t ring_size = sources[i].outputs.size();
+    rv.mixRing[i].resize(ring_size);
+    for (std::size_t j = 0; j < ring_size; ++j) {
+      rv.mixRing[i][j].dest = sources[i].outputs[j].second.dest;
+      rv.mixRing[i][j].mask = sources[i].outputs[j].second.mask;
+    }
+  }
+
+  // make pseudo-output commitments
+  rct::keyV a;  //pseudo-output commitment blinding factors
+  auto a_wiper = epee::misc_utils::create_scope_leave_handler([&]{
+    memwipe(static_cast<rct::key *>(a.data()), a.size() * sizeof(rct::key));
+  });
+  if (not reconstruction) {
+    a.resize(num_sources);
+    rv.p.pseudoOuts.resize(num_sources);
+    a[num_sources - 1] = rct::zero();
+    for (std::size_t i = 0; i < num_destinations; ++i) {
+      sc_add(
+        a[num_sources - 1].bytes,
+        a[num_sources - 1].bytes,
+        output_amount_masks[i].bytes
+      );
+    }
+    for (std::size_t i = 0; i < num_sources - 1; ++i) {
+      rct::skGen(a[i]);
```

Comment:
FWIW, I don't think the randomized values from all participants needs to influence this value, just the `s` values.

Reply:
This code block is only executed `if (not reconstruction)`, which is only true for the initial tx proposer.

Reply:
Yes, I know.

Reply:
Sorry - this comment was (somewhat) pointless, I was stating that I think all participants need to alter `s` somehow, but I didn't think these randomly generated `a` values needed that.